Adding Security Guide (#282)
nordcraftengine/documentation: docs/18-guides/02-security/index.md

---
title: App Security Guide

> description: Secure your Nordcraft app by avoiding frontend secrets and using a backend to handle sensitive API operations safely

Reviewer: ```suggestion
description: Learn how to keep your Nordcraft front-end app secure by being mindful of exposing secret API credentials and other things to avoid.
```

---

> # How to Keep Your Nordcraft App Secure

Reviewer: Please can you reformat all headings to use sentence case, e.g. "How to keep your Nordcraft app secure"

If you're building with **Nordcraft**, it's important to understand that it's a **front-end only** tool. That means everything you add—every bit of logic, config, or data—is visible to anyone using your app.

This guide covers the key things you should know (and avoid) to keep your Nordcraft app safe, especially when working with APIs or anything remotely sensitive.

---

## First Things First: What “Front-End Only” Really Means

Nordcraft runs **entirely in the browser**. There’s no backend, no secret layer, and nothing hidden from users. So when we say front-end only, we mean:

* Users can see **everything** — your UI logic, network requests, and any data you embed
* There’s **no built-in way** to protect API keys or hide sensitive operations
* It’s fantastic for rapid development and sharing ideas, but it’s not a secure place to put private stuff

---

## Don’t Store Secrets in Your App

> This is the big one. **Never** put secrets—like API keys, tokens, or credentials—directly into your Nordcraft project. If it’s in your project, it’s exposed.

Reviewer: ```suggestion
Because everything in your Nordcraft apps are visible to users in a browser, you should not store secrets or sensitive information in your Nordcraft projects.
```

### Things you absolutely shouldn’t include:

* API keys (e.g., OpenAI, Google Maps, Firebase...)
* Backend service credentials (like Supabase service roles)
* Secret tokens (OAuth, JWTs, ...)
* Private endpoints or anything involving sensitive user data

If you do add something like this by mistake, remove it immediately and rotate the key.

---

## Use a Backend to Handle Sensitive Stuff

Need to work with APIs that require secrets? Or want to protect access to certain features? That’s where a **backend** comes in. Nordcraft doesn’t stop you from building full apps—it just means **you’ll need a separate server or service to do the secure work.**

> Your backend should:

Reviewer: ```suggestion
Back end applications can more securely:
```

* Store and manage API keys
* Handle user authentication and permissions

> * Rate-limit usage and protect against abuse

Reviewer: ```suggestion
* Handle rate-limit usage and protect against abuse
```

> * Proxy API requests so the front-end never sees your keys

Reviewer: ```suggestion
* Proxy API requests, so that front-end applications do not have access to sensitive data such as API keys
```

> ### Popular options that work well with Nordcraft:

Reviewer: ```suggestion
```

| Platform               | Why use it                                        |

> | ---------------------- | ------------------------------------------------- |

Reviewer: ```suggestion
```

> | **Xano**               | Great for no-code or low-code backend needs       |

Reviewer: ```suggestion
```

> | **Supabase**           | Ideal for devs, open-source, works like Firebase  |

Reviewer: ```suggestion
```

> | **Cloudflare Workers** | Lightweight, fast, and scalable for handling APIs |

Reviewer: ```suggestion
```

> Even if your backend just forwards requests to another API, it’s a massive improvement over exposing secrets in the browser.

Reviewer: ```suggestion
```

---

> ## A Few More Tips for Security

Reviewer: ```suggestion
## Security considerations
```

Here are a few best practices that are worth following no matter what kind of app you're building:

> ### 1. Don’t trust the front-end

Reviewer: ```suggestion
### 1. Be mindful of what you include in the front end
```

> This is true for any web app, but especially here: **don’t rely on front-end logic to protect anything.** Users can bypass checks, modify code in the browser, and access anything the front-end touches.

Reviewer: ```suggestion
Anything you include in your front-end applications can be accessed in a browser. When adding any new functionality to your projects, check whether any exposed data or credentials might compromise the security of your projects.

Even seemingly benign configurations like debug flags or API base URLs can help attackers understand your project setup, which may make it vulnerable to attack.
```

Always enforce permissions and checks on the backend.

### 2. Be smart about authentication

> If your app has users, logins, or restricted features:

Reviewer: ```suggestion
If your Nordcraft application has users, logins, or restricted features:
```

* Handle sign-ins through your backend
* Use secure cookies or access tokens (preferably HTTP-only)
* Make sure every sensitive route checks permissions

> Don't let your front-end “guess” if a user is allowed to do something—have the backend confirm it.

Reviewer: ```suggestion
```

### 3. Watch your API usage

> If your app talks to an API:

Reviewer: ```suggestion
If your Nordcraft application communicates with APIs:
```

> * Make sure your backend is rate-limiting requests

Reviewer: ```suggestion
* Ensure your backend is rate-limiting requests
```

* Avoid overexposing third-party APIs in the browser
* Monitor logs for abuse if your app starts to grow

### 4. Don’t leak metadata or environment details

> Even small things like debug flags, API base URLs, or project structure can help attackers understand your setup. Keep it clean and minimal.

Reviewer: ```suggestion
```

> ---

Reviewer: ```suggestion
```

> ## What About Future Nordcraft Features?

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> Right now, Nordcraft doesn’t support features like:

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> * Secret storage

Reviewer: ```suggestion
```

> * Built-in API protection or rate-limiting

Author: not sure, just asking can something of this sort + more, go into #198 ?

> * Built-in API protection or rate-limiting

Reviewer: I think we need to discuss internally how to word this properly in a separate section, as we want to cover the whole app.

It's also better to focus on what we _can_ do right now and offer guidance, rather than scare people about what they cannot do.

> * Secure server-side logic

Reviewer: ```suggestion
```

> Those things **might** come in future updates, but for now, the safest route is to assume everything is public and handle any secure logic outside the app.

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> ---

Reviewer: ```suggestion
```

> ## Summary: What to Do and What to Avoid

Reviewer: ```suggestion
```

> | Do this                                | Avoid this                                          |

Reviewer: ```suggestion
```

> | -------------------------------------- | --------------------------------------------------- |

Reviewer: ```suggestion
```

| Use Nordcraft for front-end UI         | Putting API keys directly in your app               |
| Set up a backend to call APIs          | Authenticating or authorising only on the front-end |
| Use cookies or tokens from the backend | Sharing links to apps with secrets inside           |
| Proxy sensitive requests securely      | Assuming your app is private—it’s not               |

> ---

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

## Final Thoughts: Don’t Share Secrets. Ever.

One last reminder: **everything in a Nordcraft project is public by design.** Even if you’re just testing or sharing with teammates, treat it like it’s live and open.

> If you need to work with protected data or secure APIs, always build that logic outside the app. Nordcraft’s job is to handle the UI—and it does that really well—but security is still in your hands.

Reviewer: ```suggestion
```

---
title: How to keep your apps secure
description: Learn how to keep your Nordcraft front-end app secure by being mindful of exposing secret API credentials and other sensitive data.
---

# How to keep your Nordcraft app secure

> Nordcraft is a front-end visual framework. That means everything you add to your application including logic, configuration, or data, is visible to anyone using your app in a browser.

Reviewer 2: It sounded a bit like this was specific to Nordcraft

This guide covers the key things you should know to keep your Nordcraft projects safe and secure, especially when working with APIs or any sensitive information.

## What does front-end framework mean?

Your Nordcraft applications run in the browser. There’s no back end, no secret layer, and nothing hidden from users. In front-end applications:

> - Users can see **everything** that makes up your app, including UI logic, network requests, and any data that powers your project, regardless of whether or not that data is shown on the page.

Reviewer 3: ```suggestion
- Users can investigate/access most things that make up your app, including UI logic, network requests, and any data that powers your project, regardless of whether or not that data is shown on the page.
```

- As a result of this, you are unable to hide API keys or other sensitive credentials from your users in your Nordcraft apps.

> 

Reviewer 3: ```suggestion

:::info
For protected pages, users cannot access UI logic, network requests etc. unless they're authenticated. For unauthenticated users, they will instead be redirected by redirect rules from an API that you define in your project.
:::

```

---
title: How to keep your apps secure
description: Learn how to keep your Nordcraft front-end app secure by being mindful of exposing secret API credentials and other sensitive data.
---

# How to keep your Nordcraft app secure

Nordcraft is a front-end visual framework. That means everything you add to your application including logic, configuration, or data, is visible to anyone using your app in a browser.

This guide covers the key things you should know to keep your Nordcraft projects safe and secure, especially when working with APIs or any sensitive information.

## What does front-end framework mean?

Your Nordcraft applications run in the browser. There’s no back end, no secret layer, and nothing hidden from users. In front-end applications:

* Users can see **everything** that makes up your app, including UI logic, network requests, and any data that powers your project, regardless of whether or not that data is shown on the page.
* As a result of this, you are unable to hide API keys or other sensitive credentials from your users in your Nordcraft apps.

> 

Reviewer: ```suggestion
```

* API keys that allow you to write data, such as content management keys provided by a CMS
* Backend service credentials (like Supabase service roles)
* Secret tokens such as OAuth or JWTs tokens
* Private endpoints or anything involving sensitive user data

If you believe you have added any sensitive data or API credentials to your Nordcraft projects, remove them as soon as you can, generate new credentials, and delete the old ones.

:::info
Some API services provide **read-only** API keys to allow you to fetch data in a browser without compromising security. These types of keys are suitable to use in your Nordcraft applications. Always consult the relevant API service documentation to make sure you're using the correct read-only keys in your projects.
:::

## Use a back end to handle sensitive data and processes

If you need to work with APIs that require the use of sensitive credentials, or if you need to protect access to certain features, you'll need to use a separate back-end application.

Back-end applications can more securely:

* Store and manage API keys
* Configure user authentication and permissions

> * Handle rate-limit usage and protect against abuse

Reviewer: ```suggestion
- Handle rate-limit usage and protect against abuse
```
